### PR TITLE
Add `compare_file_screenshot()`; Add `AppDriver$expect_screenshot(compare=)`

### LINKED ISSUE #218: bs4Dash Layout Snapshot Problems
Hi there! 

Was working through building out some new testing frameworks on some of our apps and ran into some trouble developing simple snapshots. We're currently using [bs4Dash](https://rinterface.github.io/bs4Dash/index.html) for a lot of our app layouts, and it looks like there may be some compatibility issues with the snapshotting features with using this layout. Wondering if anyone has had these same issues and if it might be possible to diagnose. 

I've developed a small reprex to illustrate the issue, which can be viewed (and forked) [here](https://github.com/Martin-McCoy/shinytest2_bs4dash_issue_repo). 

In short, it appears as though there are small differences in the margins/re-sizing of elements when snapshots are taken that are generating differences between the reference images and the new snapshots. We've tried a few workarounds using `Sys.sleep()` to determine if it might be an animation wait-time issue, with little success. Any help anyone could provide would be great - thank you!

### LINKED ISSUE #220: Rounded corners are not consistent
Related #136
Related #4 
Related #218

Rounded corners are not being produced consistently on the same platforms. Shadows and borders are particularly difficult as they are being calculated with a bit of randomness. I believe this is due to inconsistent sub pixel rounding and/or snapping.

Related chromium issues: 
* Issue 937273: Different rounding for subpixel border vs outline - https://bugs.chromium.org/p/chromium/issues/detail?id=937273&q=subpixel%20snapping&can=2
* Issue 1313330: Color picker's border displays incorrectly in high contrast mode - https://bugs.chromium.org/p/chromium/issues/detail?id=1313330
* Issue 1267773: One-off pixel error on layer  - https://bugs.chromium.org/p/chromium/issues/detail?id=1267773#c9

This was originally addressed in #136 by not adjusting the screen size within `{chromote}`. While the output cards were addressed, the change in `{chromote}` did not fix _all_ rounded corner issues.

Example: 

See red corners of both the modal and focused text in the image below. Red color means that a difference was found using the command `gm compare old.png new.png -highlight-style assign -file diff.png`.

image

-----------------------------------------------------

Proposal: 

Add a `compare` function to `AppDriver$expect_sreenshot(compare=)`. Add a new comparison method that computes a "difference" metric and compare it against a set threshold.

The current compare function is set to `testthat::compare_file_binary(old, new)`. 

Proposed `shinytest2::image_threshold(threshold = 0, size = 100)` algorithm:
* Compare images using `testthat::compare_file_binary()`
	* If successful, return success
* Read in red, green, blue changes of each file
* Find abs differences in each channel for each file
* If threshold  == 0 || size <= 1
	* Return `sum(abs_diffs) == 0`
* Run convolution over diff matrix using a kernal w/ width and height set to `size`
* `total = sum(convo_vals)`
* Return `total < threshold`

-----------------------

Implementation would be done using `{cpp11}`.

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #218
Fixes #220 

~~(Creating PR now, will update docs before merge)~~; Updated docs for methods

### CODE REVIEW

**Comment by schloerke** [comment]
Some widows have a very large diff in the value found in a test

I wonder if reading a png is different in different operating systems. Or if finding the diff matrix is different on different OS values

`routine` is a bug in shiny-workflows / pak. IDK which right now

**Comment by schloerke** [comment]
Found a bug in which I was accessing matrix locations that were not initialized in cpp. 🤦‍♂️

**Review by wch — APPROVED** [review]

File: R/app-driver.R [hunk 1]
```diff
@@ -989,28 +989,104 @@ AppDriver <- R6Class( # nolint
...
+    #' )
     #' }
     expect_screenshot = function(
       ...,
+      threshold = NULL,
+      kernel_size = 5,
       screenshot_args = missing_arg(),
       delay = missing_arg(),
       selector = missing_arg(),
+      compare = missing_arg(),
```
Comment (wch) [inline R/app-driver.R:1078]: For future reference, here's another possibility for the API.

We could remove the `threshold` and `kernel_size` params from this function, and the default value for `compare` would be:

```R
  compare = testthat::compare_file_binary
```

And then if people want to use the fuzzy comparison, they would do something like:

```R
  compare = fuzzy_comparer(threshold = 3, kernel_size = 5)

  # Or maybe those values could use those as defaults
  compare = fuzzy_comparer()
```

The benefit is that it's a more general API, and provides a cleaner interface for plugging in other comparison functions. However, it seems unlikely anyone would ever use this capability, and it requires people to know about the `fuzzy_comparer` function.

**Comment by wch** [comment]
Here's a diagram illustrating the algorithm when traversing a row.

image

**Comment by schloerke** [comment]
Merging and keeping unresolved convo open as it is a comment.

**Pull request merged**